R/method-correlation.R

```diff
@@ -0,0 +1,173 @@
+    }
+
+    # Check that variables is the correct class
+    if (!is.null(variables) && !'VariableMetadataList' %in% class(variables)) {
+      stop("The variables argument must be a VariableMetadataList if defined.")
+    }
+
+    computeMessage <- ''
+    veupathUtils::logWithTime(paste("Received df table with", nrow(df1), "samples and", (ncol(df1)-1), "taxa."), verbose)
+
+    # Replace NA values with 0
+    if (naToZero) {
+      veupathUtils::setNaToZero(df1)
+      veupathUtils::logWithTime("Replaced NAs with 0.", verbose)
+    }
+
+    # Ensure the sample IDs, and their order, are the same in both datasets
+    if (!identical(df1[[recordIdColumn1]], df2[[recordIdColumn2]])) {
+      if (!identical(union(df1[[recordIdColumn1]], df2[[recordIdColumn2]]), intersect(df1[[recordIdColumn1]], df2[[recordIdColumn2]]))) {
+        stop("Sample IDs do not match between data1 and data2.")
+      }
+      # Reorder df2 (sample metadata) so that it matches the order in df
+      sampleMetadata <- df2[match(df1[[recordIdColumn1]], df2[[recordIdColumn2]]), ]
+      veupathUtils::logWithTime("Reordered sampleMetadata rows based on abundance data sample id order.", verbose)
+    }
+
+    ## Extract appropriate metadata columns
+    if (!is.null(variables)) {
+      # Use the data shape to find appropriate columns in the metadata
+      inputMetadataCols <- veupathUtils::findColNamesByPredicate(variables, function(x) {identical(x@dataShape@value, "CONTINUOUS")})
+
+      # If any are dates, we'll need to coerce them to numeric for the correlation
+      dateMetadataCols <- veupathUtils::findColNamesByPredicate(variables, function(x) {identical(x@dataShape@value, "CONTINUOUS") && identical(x@dataType@value, "DATE")})
+      if (!!length(dateMetadataCols)) {
+        veupathUtils::logWithTime("Converting date variables to numeric", verbose)
+        df2[, (dateMetadataCols) := lapply(.SD, as.numeric), .SDcols = dateMetadataCols]
+      }
+
+    } else {
+      warning("No variable metadata supplied. Using all numeric columns of sampleMetadata for correlation.")
+      inputMetadataCols <- veupathUtils::findNumericCols(df2[, -..allIdColumns2])
+    }
+    if (length(inputMetadataCols) < 1) {
+      stop("correlation requires at least one continuous metadata variable.")
+    }
+
+
+    ## Compute correlation
+    # Result has columns "data1", "data2", and correlationCoef
+    corrResult <- correlation(df1[, -..allIdColumns1], df2[, ..inputMetadataCols], method = method, verbose = verbose)
+
+
+    ## Format results
+    # Rename the correlation results columns so that they are a little more friendly
+    names(corrResult)[names(corrResult) == "data1"] <- data1Name
+    names(corrResult)[names(corrResult) == "data2"] <- data2Name
+
+    
+    # Construct the ComputeResult
+    result <- new("ComputeResult")
+    result@name <- 'correlation'
+    result@recordIdColumn <- recordIdColumn1
+    result@ancestorIdColumns <- ancestorIdColumns1
```

Comment:
are we making an assumption here that the AbundanceData is on a child node of the SampleMetadata? or can we somehow explicitly validate or compare the ancestor ids of the two data.tables?

Reply:
I need the record ids to match so the sample metadata should be already wrangled into the assay entity space. So i guess we could check to make sure the ancestor ids match? We should add a similar check to the diff abund method, then, too